smithy-model/src/main/java/software/amazon/smithy/model/shapes/SmithyIdlModelSerializer.java

```diff
@@ -0,0 +1,804 @@
+            shape.getAllTraits().values().stream()
+                    .filter(trait -> !(trait instanceof DocumentationTrait))
+                    .filter(traitFilter)
+                    .sorted(Comparator.comparing(trait -> trait.toShapeId().toString(), String.CASE_INSENSITIVE_ORDER))
+                    .forEach(this::serializeTrait);
+        }
+
+        private void serializeDocumentationTrait(DocumentationTrait trait) {
+            // The documentation trait has a special syntax, which we always want to use.
+            codeWriter.setNewlinePrefix("/// ")
+                    .write(trait.getValue().replace("$", "$$"))
+                    .setNewlinePrefix("");
+        }
+
+        private void serializeTrait(Trait trait) {
+            Node node = trait.toNode();
+            Shape shape = model.expectShape(trait.toShapeId());
+
+            if (trait instanceof BooleanTrait || isEmptyStructure(node, shape)) {
+                // Traits that inherit from BooleanTrait specifically can omit a value.
+                // Traits that are simply boolean shapes which don't implement BooleanTrait cannot.
+                // Additionally, empty structure traits can omit a value.
+                codeWriter.write("@$I", trait.toShapeId());
+            } else if (node.isObjectNode()) {
+                codeWriter.writeIndent().openBlockInline("@$I(", trait.toShapeId());
+                nodeSerializer.serializeKeyValuePairs(node.expectObjectNode(), shape);
+                codeWriter.closeBlock(")");
+            } else {
+                codeWriter.writeIndent().writeInline("@$I(", trait.toShapeId());
+                nodeSerializer.serialize(node, shape);
+                codeWriter.write(")");
+            }
+        }
+
+        private boolean isEmptyStructure(Node node, Shape shape) {
+            return !shape.isDocumentShape() && node.asObjectNode().map(ObjectNode::isEmpty).orElse(false);
+        }
+
+        @Override
+        public Void listShape(ListShape shape) {
+            shapeWithMembers(shape, Collections.singletonList(shape.getMember()));
+            return null;
+        }
+
+        @Override
+        public Void setShape(SetShape shape) {
+            shapeWithMembers(shape, Collections.singletonList(shape.getMember()));
+            return null;
+        }
+
+        @Override
+        public Void mapShape(MapShape shape) {
+            shapeWithMembers(shape, ListUtils.of(shape.getKey(), shape.getValue()));
+            return null;
+        }
+
+        @Override
+        public Void structureShape(StructureShape shape) {
+            shapeWithMembers(shape, sortMembers(shape.getAllMembers().values()));
+            return null;
+        }
+
+        @Override
+        public Void unionShape(UnionShape shape) {
+            shapeWithMembers(shape, sortMembers(shape.getAllMembers().values()));
+            return null;
+        }
+
+        private List<MemberShape> sortMembers(Collection<MemberShape> members) {
```

Comment:
Yes, but since their `compareTo` is case-sensitive then the order in the map will be too. I guess all this case-insensitivity doesn't *really* matter, but I personally hate trying to find something in a case-sensitive list.

Reply:
Case sensitivity is preferred IMO, and it would simplify some of this code. I'm not sure how an insensitive ordering makes things easier to find.

Reply:
It's not as big a deal for members, but it's a pretty hefty downside for shapes in some cases.